### PR TITLE
[9.0][ADD] Password Security Settings

### PR DESCRIPTION
**Pull request by lasley**
# 
# Password  Security

This module allows admin to set company-level password security requirements
and enforces them on the user.

It contains features such as
- Password expiration days
- Password length requirement
- Password minimum number of lowercase letters
- Password minimum number of uppercase letters
- Password minimum number of numbers
- Password minimum number of special characters
# Configuration
- Navigate to company you would like to set requirements on
- Click the `Password Policy` page
- Set the policies to your liking.

Password complexity requirements will be enforced upon next password change for
any user in that company.
## Settings & Defaults

These are defined at the company level:

| Name | Default | Description |
| --- | --- | --- |
| password_expiration | 60 | Days until passwords expire |
| password_length | 12 | Minimum number of characters in password |
| password_lower | True | Require lowercase letter in password |
| password_upper | True | Require uppercase letters in password |
| password_numeric | True | Require number in password |
| password_special | True | Require special character in password |

### CODE REVIEW

**Comment by lasley** [comment]
~~Hi @oca-clabot - @jamesfoster85 was covered under LasLabs ECLA when his commit was made~~ I squashed away from his commit anyways now 😄

**Comment by lasley** [comment]
Going to add tests to cover the controller, but would love to know anyone's thought in the interim 😄

File: res_users_password_security/README.rst
Comment (elicoidal) [inline res_users_password_security/README.rst]: One common requirement for password security is the impossibility to repeat passwords. Would it be a good idea to add (maybe in roadmap?)

File: res_users_password_security/README.rst
Comment (elicoidal) [inline res_users_password_security/README.rst]: use enumerated list for procedures ("#. ")

File: res_users_password_security/README.rst
Comment (elicoidal) [inline res_users_password_security/README.rst]: just a note: why dont use simple rst table format? much easier to handle and maintain. In your case you dont need the complex format

File: res_users_password_security/models/res_users.py
Comment (elicoidal) [inline res_users_password_security/models/res_users.py]: What if all checkboxes are set to false? you would still get the message while not necessary, right?

File: res_users_password_security/tests/company.yml
Comment (elicoidal) [inline res_users_password_security/tests/company.yml]: yml tests are deprecated. Please use py tests.

File: res_users_password_security/views/res_company_view.xml
Comment (elicoidal) [inline res_users_password_security/views/res_company_view.xml]: odoo

File: res_users_password_security/views/res_company_view.xml
Comment (elicoidal) [inline res_users_password_security/views/res_company_view.xml]: you can remove

File: res_users_password_security/README.rst
Comment (elicoidal) [inline res_users_password_security/README.rst]: be careful: be consistent in your licensing. Everywhere in your headers you state LGPL

**Comment by elicoidal** [comment]
@lasley Thanks for the module !!!!

File: res_users_password_security/README.rst
Comment (lasley) [reply res_users_password_security/README.rst]: We're talking about the reuse of old passwords right? 

Will probably need to add something to limit the amount of resets in a timespan as well then - I know that's how I get around these types of requirements 😛

Comment (elicoidal) [reply res_users_password_security/README.rst]: yes indeed.

Comment (lasley) [reply res_users_password_security/README.rst]: I gave it a valiant effort, but my strategy seems to be a bust. These `pbkdf2_*` hashes return a different value each time, similar to encryption. This defeats the normal strategy of saving and referencing the pass hashes directly for use as a history. This just blew my limited knowledge of cryptography out of the water - crazy smart people with their crazy math.

I'm open to ideas if you have them. The last commit removes my attempt if you want to see what I had.

```
>>> pbkdf2_sha512.encrypt("password")
'$pbkdf2-sha512$25000$4ByDsPa.lxJCaE2JkVLK.Q$dVDsbr49NHfKEQWYdjunuqbz2a2MGQmj0rfF3Bh519StAUouvCyYag50EL7kjBckyArqDRBdrwDHkeWeHva5DA'
>>> pbkdf2_sha512.encrypt("password")
'$pbkdf2-sha512$25000$RYixlnKOMab0vjemtNa6lw$d/8zZUv4Ttjfj31/LrXulIk8EuyPPptGpwWllm8jEtU0Ehoe7UfjPmzdn8KwpM8o6eTShCFjn7TvfFzSpf.J3Q'
>>> pbkdf2_sha512.encrypt("password")
'$pbkdf2-sha512$25000$AQBgLEVoLSVEKIUwZuzd.w$BJkX.U1YJTO5Q.g5I/1kh4guNF0qG/1/cYJ8amhFB0JSRT65oPtuJJtwGvNNUGOcdElKuNsXyyitAhcWZdK6Ng'
```

Comment (hbrunn) [reply res_users_password_security/README.rst]: modern hashing algorithms don't only depend on the input string, but also a salt (another input string) and the amount of rounds: https://pythonhosted.org/passlib/lib/passlib.hash.pbkdf2_digest.html#interface
Which is a good thing, and you shouldn't make an attempt to fix the salt or whatever. Just take the strings as they are, and try new passwords against them. The result contains those values, so you don't need to worry about it: https://pythonhosted.org/passlib/lib/passlib.hash.pbkdf2_digest.html#format-algorithm

Comment (lasley) [reply res_users_password_security/README.rst]: I guess I've only seen crap implementations like Magento then 😛 . 

This make sense though, so basically instead of trying to reference hash vs hash, I need to validate the non-hashed password against the previous X hashed. Piece of cake, thanks @hbrunn

Comment (lasley) [reply res_users_password_security/README.rst]: 👍  worked like a champ, thanks again! Implementing the minimum reset time then I think we'll be good.

File: res_users_password_security/__openerp__.py
Comment (lasley) [inline res_users_password_security/__openerp__.py]: FYI I have to leave this version number, this is already in the wild under our repos. We are migrating here and will be having our's removed from the app store for a hopefully seamless transition.

File: res_users_password_security/README.rst
Comment (elicoidal) [reply res_users_password_security/README.rst]: :+1: excellent. 
If I understand well you will have to save a history of hashes so that we make sure the user doesnot repeat passwords.

Comment (lasley) [reply res_users_password_security/README.rst]: Yup that's implemented in [`res.users.pass.history`](https://github.com/laslabs/server-tools/blob/9d20afcbaebc520818ebb8f9462d7932ebe5c48d/res_users_password_security/models/res_users_pass_history.py#L8) 😄

File: res_users_password_security/__openerp__.py
Comment (elicoidal) [reply res_users_password_security/__openerp__.py]: let's say this is not yet critical until merge. After the merge we should stick strictly to the rule

Comment (lasley) [reply res_users_password_security/__openerp__.py]: Once merged, people that are using the `9.0.1.0.1` version from my repo will need a way to indicate the change won't they? I honestly have no idea how the Odoo App store works, but I figured these version numbers has some relevance to the auto update?

Comment (elicoidal) [reply res_users_password_security/__openerp__.py]: My point is follow your internal logic as this is not important at which number you start in the OCA repo. What is important though is that after the merge the numbers are dealt logically

File: res_users_password_security/README.rst
Comment (elicoidal) [reply res_users_password_security/README.rst]: Great :)

File: res_users_password_security/__openerp__.py
Comment (lasley) [reply res_users_password_security/__openerp__.py]: Ahh cool thanks for elaborating 👍

**Comment by lasley** [comment]
Alright assuming this build goes green, I'm done here 😄 

Will squash my commits once we get the thumbs

File: res_users_password_security/models/res_company.py
Comment (elicoidal) [inline res_users_password_security/models/res_company.py]: "and 0 to disable"

File: res_users_password_security/models/res_users.py
Comment (elicoidal) [inline res_users_password_security/models/res_users.py]: don't you overwrite the message variable here?

**Comment by lasley** [comment]
Thanks for the catches @elicoidal - handled

**Comment by dreispt** [comment]
AFAICS we already have one (or more) PR for this feature.
Did you check #294 ?

**Comment by lasley** [comment]
@dreispt - I was unaware of that PR, but there are too many issues with it in my opinion & it doesn't seem like it is being worked. 

Main non-starter in my eyes is the breaking of the inheritance chain, massive code duplication & 0% test coverage. 

You're welcome to block this PR in favor of that if you would like, but unfortunately I will not be fixing that one as there are more features here, it is tested, and we are not breaking inheritance chains.

**Comment by dreispt** [comment]
I don't want to block anything, we just need to avoid duplicate PRs.
The same story happened with the previous one - there were two different PRs for similar features.
I involved both authors in a discussion of the pros and cons of each, and finally the most promising PR was selected and the other one was closed.

So, we need to invite the other PR's author(s) here, compare the features here with the referenced PR. If  the other PR won't go forward, at least there may be good ideas there that could be picked up here, and we could ensure that this PR also fits the use cases for the other PR.

**Comment by lasley** [comment]
Cool I understand the motivation, thanks for elaborating @dreispt

@hhgabelgaard - I seem to have duplicated your functionality from #294. I propose that we merge this PR instead, then backport to v9 in order to provide the same features of your PR plus
- Handling for lowercase/uppercase rules
- Minimum password length (aka user cannot reset for 24 hours since last reset)
- Disallow use of X last passwords
- Password expiration

File: res_users_password_security/views/res_company_view.xml
Comment (bealdav) [inline res_users_password_security/views/res_company_view.xml]: May be a module you don't know could have been used here.
https://github.com/OCA/server-tools/tree/9.0/configuration_helper
It allow to define settings easily on settings page.
Non blocking for me

File: res_users_password_security/README.rst
Comment (bealdav) [inline res_users_password_security/README.rst]: I think module technical name could be 'password_security'. 
Display name and technical name should be the same (except plural maybe), I think at least.
2 words is quickly read, and when we read password, everybody think about users.

**Comment by bealdav** [comment]
I have tested this code with V8. It works, except <odoo> instead of <openerp>  in xml files

Thanks for this really good job fully tested.
:+1: 

@hhgabelgaard are you ok to close your PR  #294 and vote +1 here

**Review by bealdav — APPROVED** [review]

File: res_users_password_security/controllers/main.py
Comment (moylop260) [inline res_users_password_security/controllers/main.py]: Could you use `except PassError as e:` instead, please?

Comment (moylop260) [reply res_users_password_security/controllers/main.py]: More info: https://www.python.org/dev/peps/pep-3110

**Comment by hhgabelgaard** [comment]
I agree, this seems better than #294 
+1

File: res_users_password_security/controllers/main.py
Comment (moylop260) [inline res_users_password_security/controllers/main.py]: Could you use `limit=1`, please?

File: res_users_password_security/controllers/main.py
Comment (moylop260) [inline res_users_password_security/controllers/main.py]: Could you use `limit=1`, please?

File: res_users_password_security/controllers/main.py
Comment (moylop260) [inline res_users_password_security/controllers/main.py]: With `limit=1` and new api you don't need a `if` validation here.

``` diff
user_ids = request.env.sudo().search(... limit=1)
...
- if len(user_ids) == 1:
-    user_ids[0]._validate_pass_reset()
+ user_ids._validate_pass_reset()
```

File: res_users_password_security/models/res_users_pass_history.py
Comment (moylop260) [inline res_users_password_security/models/res_users_pass_history.py]: If we delete a user: Should we store the pass history?
If the answer is `no`: you could use `ondelete=cascade`

**Review by moylop260 — CHANGES_REQUESTED** [review]

File: res_users_password_security/models/res_users.py
Comment (moylop260) [inline res_users_password_security/models/res_users.py]: We don't need the parens here
`return days > self.company_id.password_expiration`

File: res_users_password_security/tests/test_password_security_home.py
Comment (moylop260) [inline res_users_password_security/tests/test_password_security_home.py]: for `__new__` method use `cls` instead of `self`

File: res_users_password_security/tests/test_password_security_home.py
Comment (moylop260) [inline res_users_password_security/tests/test_password_security_home.py]: What about use a name of variable different with `snake_case` for `Controller`?

File: res_users_password_security/tests/test_password_security_session.py
Comment (moylop260) [inline res_users_password_security/tests/test_password_security_session.py]: What about use a name of variable different with `snake_case` for `Controller`?

**Comment by lasley** [comment]
Thanks for the reviews @bealdav & @moylop260 - all changes made outside of the `configuration_helper`. You are correct that I was unaware of the existence of this module, but will need to play around with it a bit first I think.

**Review by moylop260 — APPROVED** [review]
Thanks

**Comment by moylop260** [comment]
Could you review the red build of travis related with this module, please?
https://travis-ci.org/OCA/server-tools/jobs/162977224

**Comment by lasley** [comment]
Alright we should be good on Travis and your last comments @moylop260

**Comment by moylop260** [comment]
ups lint show

``` bash
password_security/tests/test_password_security_home.py:235:17: F841 local variable 'search' is assigned to but never used
password_security/tests/test_password_security_home.py:264:17: F841 local variable 'search' is assigned to but never used
password_security/tests/test_password_security_home.py:266:17: F841 local variable 'user' is assigned to but never used
```

**Comment by lasley** [comment]
Woot build 🍏

**Comment by bealdav** [comment]
:+1: thanks

**Comment by dreispt** [comment]
🎉

**Pull request merged**
